### PR TITLE
Wcs from points

### PR DESCRIPTION
**Pull request by shannnnnyyy**
added wcs_from_points function in wcs.utils. closes #7884 (opened new PR because rebase failed)

### CODE REVIEW

File: astropy/wcs/utils.py [hunk 1]
```diff
@@ -14,13 +13,12 @@
            'proj_plane_pixel_area', 'is_proj_plane_distorted',
            'non_celestial_pixel_scales', 'skycoord_to_pixel',
            'pixel_to_skycoord', 'custom_wcs_to_frame_mappings',
-           'custom_frame_to_wcs_mappings', 'pixel_to_pixel']
+           'custom_frame_to_wcs_mappings', 'fit_wcs_from_points']
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: This looks like a rebasing issue, ``pixel_to_pixel`` should not be removed from this list

File: astropy/wcs/utils.py [hunk 2]
```diff
@@ -14,13 +13,12 @@
...
 
 
 def add_stokes_axis_to_wcs(wcs, add_before_ind):
     """
     Add a new Stokes axis that is uncorrelated with any other axes.
-
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: Please revert these changes to the whitespace here and below - we like having the empty lines for readability :)

File: astropy/wcs/utils.py [hunk 3]
```diff
@@ -635,193 +585,225 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
...
-def _unique_with_order_preserved(items):
-    """
-    Return a list of unique items in the list provided, preserving the order
-    in which they are found.
-    """
-    new_items = []
-    for item in items:
-        if item not in new_items:
-            new_items.append(item)
-    return new_items
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: This code should not be removed, and indicates a rebasing issue

File: astropy/wcs/utils.py [hunk 4]
```diff
@@ -635,193 +585,225 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
...
-        world_indices = list(np.nonzero(world_include)[0])
-        pixel_used.extend(pixel_indices)
-        split_info.append((pixel_indices, world_indices))
 
-    return split_info
+def fit_wcs_from_points(xy, world_coords, proj_point='center', 
+                        projection='TAN', sip_distortion=True, degree=4):
 
+    """ 
+    Given two matching sets of coordinates on detector and sky,
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: The first line of the docstring should give a one sentence summary, then empty line, then full description.

File: astropy/wcs/utils.py [hunk 5]
```diff
@@ -635,193 +585,225 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
...
+        b_vals = np.zeros((degree+1, degree+1))
+
+        for coef_name in coef_names:
+            a_vals[int(coef_name[0])][int(coef_name[2])] = coef_fit[0].pop(0)
+            b_vals[int(coef_name[0])][int(coef_name[2])] = coef_fit[1].pop(0)
+
+        wcs.sip = Sip(a_vals, b_vals, np.zeros((degree+1, degree+1)), 
+                      np.zeros((degree+1, degree+1)), wcs.wcs.crpix) 
+
+    return wcs
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: Add an empty line at the end of the file

File: astropy/wcs/utils.py [hunk 6]
```diff
@@ -635,193 +585,225 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
...
+                         "the polynomial order must be provided.")
+
+    # set pixel_shape to span of input points
+    wcs.pixel_shape = (max(xp)-min(xp), max(yp)-min(yp))
+
+    # determine CRVAL from input
+    close = lambda l, p: p[np.where(np.abs(l) == min(np.abs(l)))[0][0]]
+    if str(proj_point) == 'center':  # use center of input points
+        wcs.wcs.crval = ((max(lon)+min(lon))/2., (max(lat)+min(lat))/2.)
+        wcs.wcs.crpix = ((max(xp)+min(xp))/2., (max(yp)+min(yp))/2.)
```
Comment (ceb8) [inline astropy/wcs/utils.py]: This will fail for a footprint that crosses the 360->0 degree crossover. In this case the CRVAL lon ends up being near 175 deg, and the wcs fitting never really recovers. This limitation should maybe be mentioned here in the doc string?

Comment (shannnnnyyy) [reply astropy/wcs/utils.py]: if i can't figure out a fix for this today ill just make a note of it in the docstring. let me know if you have any ideas - i'll see if I can use something in coordinates to generalize how the center of points is computed

Comment (eteq) [reply astropy/wcs/utils.py]: I'm ok with this being a "known issue" to fix later, but if possible it would be nice to add a test case that covers this but does @xfail, and then we can make an issue in the tracker that's basically just "make the xfail not fail".

Comment (ceb8) [reply astropy/wcs/utils.py]: Yeah, I couldn't come up with a good fix for it either when I encountered it. I ended up just avoiding using 'center' if I didn't know where a footprint was going to land.

Comment (shannnnnyyy) [reply astropy/wcs/utils.py]: issue #5766 has a solution for this, testing it now

Comment (shannnnnyyy) [reply astropy/wcs/utils.py]: maybe you could test this on one of the cases that was failing before?

File: astropy/wcs/tests/test_utils.py [hunk 7]
```diff
@@ -990,3 +994,77 @@ def test_pixel_to_pixel_1d():
...
+
+    # Calculating the true sku positions
+    world_pix = SkyCoord(full_wcs.all_pix2world(list(zip(x,y)), 1), unit='deg')
+
+    # Fitting the wcs
+    linear_wcs = fit_wcs_from_points((x, y), world_pix,
+                                     proj_point='center', sip_distortion=False)
+
+    # Getting the sky positions using the fitted wcs
+    world_pix_new = SkyCoord(linear_wcs.all_pix2world(list(zip(x,y)), 1), unit='deg')
```
Comment (astrofrog) [inline astropy/wcs/tests/test_utils.py]: As above you can do:

```
world_pix_new = linear_wcs.pixel_to_world(x, y)
```

File: astropy/wcs/utils.py [hunk 8]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
                 outputs[ipix] = np.broadcast_to(pixel_outputs[ipix], original_shape)
 
     return outputs[0] if wcs_out.pixel_n_dim == 1 else outputs
+
+def _linear_wcs_fit(params, lon, lat, x, y, w_obj):
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: Add an empty line above to avoid PEP8 issues

File: astropy/wcs/utils.py [hunk 9]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+    w_obj.wcs.cd = ((cd[0], cd[1]), (cd[2], cd[3]))
+    w_obj.wcs.crpix = crpix
+    lon2, lat2 = w_obj.wcs_pix2world(x, y, 1)
+    resids = np.concatenate((lon-lon2, lat-lat2))
+
+    return resids
+    
+
+def _sip_fit(params, lon, lat, u, v, w_obj, order, coeff_names):
+
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: Remove this empty line

File: astropy/wcs/utils.py [hunk 10]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
                 outputs[ipix] = np.broadcast_to(pixel_outputs[ipix], original_shape)
 
     return outputs[0] if wcs_out.pixel_n_dim == 1 else outputs
+
+def _linear_wcs_fit(params, lon, lat, x, y, w_obj):
+    """Objective function for fitting linear terms.
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: Add newline after """ and add empty line after 'terms.'

File: astropy/wcs/utils.py [hunk 11]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+        WCS object
+    """
+
+    from ..modeling.models import SIP, InverseSIP   # here to avoid circular import
+
+    # unpack params
+    crpix = params[0:2]
+    cdx = params[2:6].reshape((2, 2))
+    a_params = params[6:6+len(coeff_names)]
+    b_params = params[6+len(coeff_names):]
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: The docstring above suggests there are only 6 elements, but this suggests otherwise?

File: astropy/wcs/utils.py [hunk 12]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+    resids[resids > 180] = 360 - resids[resids > 180]
+    resids[resids < -180] = 360 + resids[resids < -180]
+
+    return resids
+
+
+def fit_wcs_from_points(xy, world_coords, proj_point='center', 
+                        projection='TAN', sip_degree=4):
+
+    """
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: Remove empty space above

File: astropy/wcs/utils.py [hunk 13]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+    resids = np.concatenate((x-xo, y-yo))
+    # to avoid bad restuls if near 360 -> 0 degree crossover
+    resids[resids > 180] = 360 - resids[resids > 180]
+    resids[resids < -180] = 360 + resids[resids < -180]
+
+    return resids
+
+
+def fit_wcs_from_points(xy, world_coords, proj_point='center', 
+                        projection='TAN', sip_degree=4):
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: I think the default should probably not include SIP distortions?

File: astropy/wcs/utils.py [hunk 14]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+    -------
+    wcs : `~astropy.wcs.WCS`
+        The best-fit WCS to the points given.
+    """
+
+    from astropy.coordinates import SkyCoord # here to avoid circular import
+    import astropy.units as u
+    from scipy.optimize import least_squares
+    from .wcs import Sip
+    import copy
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: The ``.copy`` and ``scipy`` imports can probably go at the top of the file

File: astropy/wcs/utils.py [hunk 15]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+
+    xp, yp = xy
+    lon, lat = world_coords.data.lon.deg, world_coords.data.lat.deg
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: I don't think you can assume that these will always use a spherical representation - maybe you put this in a ``try...except`` and if an ``AttributeError`` is raised you can access:

```
unit_sph =  world_coords.unit_spherical
lon, lat = unit_sph.lon.deg, unit_sph.lat.deg
```

File: astropy/wcs/utils.py [hunk 16]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+        Defaults to 'center', in which the geometric center of input world
+        coordinates will be used as the projection point. To specify an exact
+        point for the projection, a Skycoord object with a coordinate pair can
+        be passed in. For consistency, the units and frame of these coordinates
+        will be transformed to match 'world_coords' if they don't.
+    projection : str or `~astropy.wcs.WCS`
+        Three letter projection code, of any of standard projections defined 
+        in the FITS WCS standard. Optionally, a WCS object with projection 
+        keywords set may be passed in. 
+    sip_degree : int
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: ``None`` or int

(and then can default to ``None``)

File: astropy/wcs/utils.py [hunk 17]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+    # Change PC to CD, since cdelt will be set to 1
+    if wcs.wcs.has_pc():
+        wcs.wcs.cd = wcs.wcs.pc
+        wcs.wcs.__delattr__('pc')
+
+    if (type(sip_degree) != type(None)) and (type(sip_degree) != int):
+       raise ValueError("sip_degree must be None, or integer.")
+
+    # set pixel_shape to span of input points
+    wcs.pixel_shape = (max(xp)-min(xp), max(yp)-min(yp))
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: If ``xp`` and ``yp`` are arrays, use ``xp.max()`` etc.

File: astropy/wcs/utils.py [hunk 18]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+
+    # determine CRVAL from input
+    close = lambda l, p: p[np.where(np.abs(l) == min(np.abs(l)))[0][0]]
```
Comment (astrofrog) [inline astropy/wcs/utils.py]:
```
np.where(np.abs(l) == min(np.abs(l))
```

can be better written (I think) as:

```
np.argmin(np.abs(l))
```

File: astropy/wcs/utils.py [hunk 19]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+    if str(proj_point) == 'center':  # use center of input points
+        sc1 = SkyCoord(min(lon)*u.deg, max(lat)*u.deg)
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: As before, use ``.min()`` and ``.max()`` if these are arrays (comment applies to other instances)

File: astropy/wcs/utils.py [hunk 20]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+    # and cd terms are way off. 
+    p0 = np.concatenate([[1., 0., 0., 1], wcs.wcs.crpix.flatten()])
+    fit = least_squares(_linear_wcs_fit, p0,
+                        args=(lon, lat, xp, yp, wcs))
+    wcs.wcs.crpix = np.array(fit.x[4:6])
+    wcs.wcs.cd = np.array(fit.x[0:4].reshape((2, 2)))
+
+    # fit SIP, if specified. Only fit forward coefficients
+    if sip_degree:
+        degree = sip_degree
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: This doesn't save a lot of characters, so might not be worth it?

File: astropy/wcs/utils.py [hunk 21]
```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
...
+                                     projection=projection)
+    else: #if projection is not string, should be wcs object. use as template.
+        wcs = copy.deepcopy(projection)
+        wcs.cdelt = (1., 1.) # make sure cdelt is 1
+        wcs.sip = None
+
+    # Change PC to CD, since cdelt will be set to 1
+    if wcs.wcs.has_pc():
+        wcs.wcs.cd = wcs.wcs.pc
+        wcs.wcs.__delattr__('pc')
```
Comment (nden) [inline astropy/wcs/utils.py]: Internally wcslib works with the PC, regardless of whether the input has CD. I wouldn't try to delete it (it probably won't anyway). When you assign to wcs.cd the flag is set automatically so that has_cd() returns True and has_pc() returns False. The above three lines are not necessary.

Comment (shannnnnyyy) [reply astropy/wcs/utils.py]: There was some discussion previously about changing the output to only have CD, since it was returning PC before. Should I change it back?

File: astropy/wcs/utils.py (hunk 11 above)
Comment (shannnnnyyy) [reply astropy/wcs/utils.py]: correct, ill fix the docs. meant to say its at LEAST 6 elements, it depends on the degree of polynomial being fit.

File: astropy/wcs/utils.py (hunk 21 above)
Comment (eteq) [reply astropy/wcs/utils.py]: I think right now it's not important how consistent the output is on this.  So as long as the resulting wcs is right, we can deal with whether there should be a cd/pc cleanup step later (which I *think* means you can delete this based on what @nden said, but hopefully the tests will let you check it)

File: astropy/wcs/utils.py (hunk 18 above)
Comment (shannnnnyyy) [reply astropy/wcs/utils.py]: this gives an error - i'll replace this if i have time

File: astropy/wcs/utils.py (hunk 21 above)
Comment (shannnnnyyy) [reply astropy/wcs/utils.py]: so we don't care if the output has a PC, even though cdelt will always be 1? or we dont care if it has a cd AND a pc (which should be the same, so just redundant info) ?

File: astropy/wcs/utils.py (hunk 18 above)
Comment (astrofrog) [reply astropy/wcs/utils.py]: ah sorry you also need to remove the ``[0][0]``, so:

```
close = lambda l, p: p[np.argmin(np.abs(l))]
```

**Comment by shannnnnyyy** [comment]
The commit history has me deleting then adding in pixel_to_pixel, and I'm not sure how to undo this, but it seems like everything else is good now.

**Comment by bsipocz** [comment]
@shannnnnyyy - I'm happy to fix that for this PR when everything else is done.

File: astropy/wcs/utils.py (hunk 21 above)
Comment (nden) [reply astropy/wcs/utils.py]: Assign to the CD as before but don't delete the PC. I don't think it deletes it anyway and may cause only harm. When you assign `wcs.wcs.cd=pc` the PC is assigned it to CD and `wcs.wcs.has_cd()` returns `True`. However internally the PC is also set to that value and the library always uses the PC. This is just a feature of the internal library. So don't delete the PC.

Comment (shannnnnyyy) [reply astropy/wcs/utils.py]: Ok I changed that - I assigned the object a CD that is the same as it's PC. However, when you print out the contents of the object it displays the PC by default, so it's not clear that this even has a CD so it seems useless to do that. Is there necessarily an issue with an object not having a PC, if it will be created from the CD internally anyway? When you construct a WCS from fits data that only has CD keywords, has_pc returns False and everything works fine. I just want to make sure because there was a change request to have the output have CD instead of PC. It could also complicate things more to have both - if the user wanted to manually reset a value in the object that has PC and CD, and they do it to just CD, the changes won't be reflected.

File: astropy/wcs/utils.py (hunk 11 above)
Comment (astrofrog) [reply astropy/wcs/utils.py]: Ok sounds good!

**Comment by eteq** [comment]
(it looked to me like the test failures were due to just some missing imports and alast minute change I asked for above, so I figured I'd push up a simple fix to see if that does it)

**Comment by eteq** [comment]
(and those tests revealed something that looked by an off-by-1-pixel error...)

**Comment by ceb8** [comment]
@bsipocz @eteq I belive I have fixed the history, squashed the commits with the accidental deletion and reintroduction of the pixel_to_pixel and related code, and then rebased onto master. That version is here: https://github.com/ceb8/astropy/tree/wcs_from_points_hist_fix

File: astropy/wcs/utils.py [hunk 22]
```diff
@@ -4,6 +4,8 @@
 
 from astropy import units as u
 from astropy.utils.misc import unbroadcast
+import copy
+from scipy.optimize import least_squares
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: This introduces a non-optional dependency on the scipy package. Can you move the import into the function below?

File: astropy/wcs/utils.py (hunk 21 above)
Comment (nden) [reply astropy/wcs/utils.py]: wcslib works internally with the PC. Try this to see what I mean:
```
w = wcs.WCS()
w.wcs.cd=[[2,0],[0,2]]
print(w.wcs)
```
This prints the values of the wcslib struct and the PC is set there. It's probably OK to delete it as you do because I don't think it deletes it from the struct but I'm not sure if it has undesirable side effects. The safest thing to do is to create a new `WCS()` object and populate it from the user supplied one. But of course this defeats the purpose of supplying one as an argument because then you have to find what the projection parameters are and how they are stored in the header.

If I am reading the code correctly these assignment does not matter. The linear fit (line 1067) does not use these values as initial values, it uses the identity matrix. So essentially the only thing that is used from the supplied wcs is the projection. Unless I'm missing something. This doesn't seem like a very robust approach in the general case. But it will probably work. Sorry I'm not giving a concrete suggestion. I would probably not do it this way and would probably put some more burden on the user to specify the projection type. Since everyone else feels like this is the way to go once the other commentas are addressed this should probably be merged.

Comment (eteq) [reply astropy/wcs/utils.py]: @nden - to be clear on this 

> So essentially the only thing that is used from the supplied wcs is the projection

I think the other thing it's doing is that it is *also* using some of the random other parameters like `LONPOLE` if they are specified right?  That was the use case I was thinking of for why you'd pass in your own WCS.

Comment (nden) [reply astropy/wcs/utils.py]: @eteq I believe `LONPOLE` is determined internally by the type of projection.  `LATPOLE` OTH may be set, though that's rare. Anyway I don't want to hold this up, when tests pass it should be merged.

Comment (shannnnnyyy) [reply astropy/wcs/utils.py]: If 'projection' is passed in as a WCS, it will also reflect any non polynomial distortion in there in the fit, which may be useful in some cases. I purposefully don't have it use CD/PC from the input WCS as an initial guess, because if the refitting is being done because something is way off then the fit might be poor. I am going to go ahead and change this though, so that it does use input CD/PC as an initial guess, and i will just put a warning in the docstring of this potential behavior.  

I agree, I think I would prefer in general to have the user specify the projection with keywords alone since there is the already the function `celestial_frame_to_wcs` that should set up a WCS with all the correct keywords based on the projection type, and NOT have an input wcs as an option for projection. However, this still works and I think we should just go ahead with this since its late in the game and I'll just fix the above issue and the pep8 stuff.

Comment (eteq) [reply astropy/wcs/utils.py]: That said I agree it would be good to be able to set the initial value!

**Comment by eteq** [comment]
(I don't really understand the codecov result, but I think it's a red herring... at any rate the core points are covered as per @astrofrog's review)

**Comment by astrofrog** [comment]
I'll go ahead and merge - thanks @shannnnnyyy, @ceb8 and @eteq!

The codecov failure is not a red herring, but an indication that not enough of the new code is covered by tests, which is what I meant before in my review. But in the interest of feature freeze, I'll merge this and it would be great if someone would be willing to contribute more tests for this function to make sure it has good coverage.

**Comment by pllim** [comment]
Re: coverage -- The coverage report on `master` with this patch is at https://codecov.io/gh/astropy/astropy/tree/8511aa6aa27d92f58d85410564b9e3c1ba89e724/astropy/wcs . Unfortunately, I am forbidden to see the line-by-line report. If you can, that's great. If not, you need to run the coverage test locally to see it. From the stats that I can see, 37 lines are not covered, but I don't know which 37.
